### PR TITLE
migrate commands to command palette (Closes #389)

### LINKED ISSUE #389: Migrate all command palette commands to Roam's new system
This allows for all command palette commands to be keyboard customizable by the end user.

Some questions to answer:
- Do we simply remove the hot_keys module in workbench in favor of this new system? I lean yes.
- We should explore migrating command palette invocations in the other major extensions as well

### PR DESCRIPTION
**Pull request by mdroidian**
Let me know if I'm on the right track 👍 @dvargas92495 

components PR: https://github.com/dvargas92495/roamjs-components/pull/3

### CODE REVIEW

**Comment by mdroidian** [comment]
Also, the [docs](https://roamresearch.com/#/app/developer-documentation/page/HlEqgAw3H) seem to strongly recommend against setting a default shortcut.
Do we still want to go ahead with that?

File: src/features/tutorials.tsx
Comment (dvargasfuertes) [inline src/features/tutorials.tsx]: instead of exporting everything out to `src/index.ts`, let's pass in `extensionAPI` into `toggleFeature` as that will limit the refactor. This also allows us to remove when a feature is toggled off

**Review by dvargasfuertes — COMMENTED** [review]
We should also make sure the `removeCommand` is added in `roamjs-components`

File: src/features/tutorials.tsx
Comment (mdroidian) [reply src/features/tutorials.tsx]: I'm pretty blocked on this.

Normally I would think to `import` from a defined `export` somewhere, but it seems `extensionAPI` is in a return as a parameter of `onload`, and I can't seem to figure out what to do with that.

Would you be able to get on a call?

Comment (mdroidian) [reply src/features/tutorials.tsx]: I think I got it.

File: src/index.ts
Comment (dvargasfuertes) [inline src/index.ts:166]: yup this is what I had in mind 👍

**Comment by mdroidian** [comment]
Struggling with Typescript definition and console error with the `decorators.tsx`: https://github.com/dvargas92495/roamjs-workbench/pull/391/commits/d4d78f7a5fb634242100410e54ea274e02348787

Typescript error:
```
Type '{ settings: 
{ get: (k: string) => unknown; getAll: () => Record<string, unknown>; 
panel: { create: (c: PanelConfig) => void; }; 
set: (k: string, v: unknown) => Promise<void>; }; 
ui: { commandPalette?: { ...; }; }; }' 
is not assignable to type 'never'.ts(2322)
```

Console error:
```
Uncaught TypeError: parseRoamMarked is not a function
    at callback (decorators.tsx:482:27)
    at callback (createHashtagObserver.js:12:13)
    at Array.forEach (<anonymous>)
    at createHTMLObserver (createHTMLObserver.js:8:27)
    at createHashtagObserver (createHashtagObserver.js:5:94)
    at toggleFeature (decorators.tsx:473:119)
    at onClick (decorators.tsx:335:7)
...
```

File: src/features/decorators.tsx
Comment (dvargasfuertes) [inline src/features/decorators.tsx]: We'll need to fix the type here: `RoamOverlayProps<{ extensionAPI: OnloadArgs["extensionAPI"]}>`

**Comment by mdroidian** [comment]
> Console error:
> 
> ```
> Uncaught TypeError: parseRoamMarked is not a function
>     at callback (decorators.tsx:482:27)
>     at callback (createHashtagObserver.js:12:13)
>     at Array.forEach (<anonymous>)
>     at createHTMLObserver (createHTMLObserver.js:8:27)
>     at createHashtagObserver (createHashtagObserver.js:5:94)
>     at toggleFeature (decorators.tsx:473:119)
>     at onClick (decorators.tsx:335:7)
> ...
> ```

Maybe this isn't actually from the code change.  Looks like something similar exists in the current version.

```
helpers.ts:120 Uncaught TypeError: i is not a function
    at callback (6a693c5a-69ce-4ef4-842d-c096939dbf20:2:1646319)
    at callback (6a693c5a-69ce-4ef4-842d-c096939dbf20:2:1245300)
    at Array.forEach (<anonymous>)
    at t.default (6a693c5a-69ce-4ef4-842d-c096939dbf20:2:1244712)
    at t.Z (6a693c5a-69ce-4ef4-842d-c096939dbf20:2:1245189)
    at _n (6a693c5a-69ce-4ef4-842d-c096939dbf20:2:1646081)
    at onClick (6a693c5a-69ce-4ef4-842d-c096939dbf20:2:1644114)
...
```

Steps to replicate:
1. Turn off all decorations
2. Disable Workbench
3. Reload Graph
4. Enable Workbench
5. Toggle all decorations.  Save.

**Comment by dvargasfuertes** [comment]
Yea this looks like a current bug from when the context decorator is enabled - I'd try to ignore this for now as its a separate issue

File: src/features/decorators.tsx
Comment (mdroidian) [reply src/features/decorators.tsx]: I went a different route and removed the `addCommand` from `toggleFeature`
It made more sense to me to not "turn on and off" the feature every decoration save.
It also removes the need for type updates.  And this makes more sense to me as well, as `extensionAPI` shouldn't be required for `Decorations`
https://github.com/dvargas92495/roamjs-workbench/pull/391/commits/3eb2707ccff6a2f399f911e118dd048ae05daeab#diff-ea4e85188fe94af5ae70fed25db2d64551b1b568075d213927ed8f69a97fb202

File: src/features/deepnav.tsx
Comment (mdroidian) [inline src/features/deepnav.tsx:876]: I'm struggling with replicating this behavior via the addCommand callback.

In the static hotkey version, this is handled with an even listener to grab the DOM element to `.blur()` on.

How can I grab and pass the currently selected DOM element to the addCommand callback?

Comment (mdroidian) [reply src/features/deepnav.tsx:876]:
```
const callback = (
  e: Event
  ) => {
  console.log(`e: ${e}`);
  }
```
This just returns an array with the block reference or null.

Any suggestions on what direction to take @dvargas92495 ?

Comment (dvargasfuertes) [reply src/features/deepnav.tsx:876]: You could use `document.activeElement` and check to see if it is a textarea

Comment (mdroidian) [reply src/features/deepnav.tsx:876]: Ah!  Nice.

I was working on something like:

```
const getHTMLElementFromUid = (uid: string) => {
  const element = document.querySelector(`[id*="${uid}"]`) as HTMLElement;
  ...
```

Comment (dvargasfuertes) [reply src/features/deepnav.tsx:876]: That could get you in trouble - the same block could be on the screen multiple times (main page + sidebar)

**Comment by mdroidian** [comment]
for https://github.com/dvargas92495/roamjs-workbench/pull/391/commits/b21a05201d06598eddc52e26ad45467fc647fdd6

Removed the toggle right/left sidebar, as those are built into Roam now.

Also, let me know how you feel about code formatting
image

I went with that because if felt more readable to me than the 200+ lines from the Prettier extension.

**Comment by mdroidian** [comment]
Issues found with current jumpNav (Hot Keys)

- `X` - Expand/Collapse block tree to a certain level, specified by the following numeric key press
    - number press adds number to block

And these appear to not work:
- `S` - Add shortcut to page to left sidebar
- `z` - Toggle Unlinked Refs
- `y` - Toggle Queries
- `b` - Jump to the bottom of page
  - jumps to top of page
-  `q` - Toggle View type
   - toggles to document, but not back to bullet

**Comment by mdroidian** [comment]
Well here's an unfortunate side effect of `addCommand`
When a command is invoke that uses the command palette sets focus of the block, then the cursor disappears.
Using a hot key, this problem does not occur.

Added timeout here: https://github.com/dvargas92495/roamjs-workbench/pull/391/commits/4629860ad8e6ec4911259fe7c6175b1c3446b16c

But I feel it is bordering an unreasonable length of time.

**Comment by dvargasfuertes** [comment]
> I went with that because if felt more readable to me than the 200+ lines from the Prettier extension.

The useful part of the prettier extension is the standardization, not necessarily the decisions it comes with by default. The problem with making exceptions is that someone else working on the team will eventually edit the file, hit save, then prettier will auto format.

If you disagree with prettier defaults, the scalable solution would be to change the prettier rules, so that it gets formatted automatically. If it's unable to be encoded, then the opinion is subjective, which means it will inevitably be overwritten by somebody else's subjective experience, and now time on the PR is being spent on code formatting instead of code content.

> Added timeout here

Man the problem here is I think this API from Roam is actually not very reliable - ideally we don't need a setTimeout at all. This should be fine for now

Gonna start reviewing PR now

File: src/index.ts
Comment (dvargasfuertes) [inline src/index.ts:179]: One meta comment for future PRs - Big refactors like this are more digestable for reviewers as small as possible. This also helps on the engineering front as well.

I still struggle with this mightily (just look at some SamePage commits). But we should always ask ourselves: What is the smallest PR that we could submit that accomplishes our goal?

For example, PR 1 could have just been this file + 1 module. The rest of the modules will ignore the param and we would be able to iterate on convention before applying it to the rest of the modules

File: src/features/alert.tsx
Comment (dvargasfuertes) [inline src/features/alert.tsx:270]: No need to change for this PR, but a note for future reference.

In typescript, when there's more than one parameter I often prefer a single object instead of multiple params. So for example, I probably would have done:

```typescript
addCommand(
        {
          label: "Create New Alert",
          callback: createNewAlert,
          extensionAPI
        },
)
```

The reason being is life becomes way easier for consumers:
- Order of parameters no longer matters
- We could freely make any of the parameters optional
- Easier to refactor in the future: adding/removing parameters doesn't involve changing all consumers in most cases
- Parameters are named, giving the consumer extra context on what the function expects

File: src/features/article.tsx
Comment (dvargasfuertes) [inline src/features/article.tsx:329]: are we passing in the ALT+SHIFT+I hotkey from below?

File: src/features/dailyNotesPopup.tsx
Comment (dvargasfuertes) [inline src/features/dailyNotesPopup.tsx:638]: Same with these, seems like we're missing the hot key for each

File: src/features/deepnav.tsx
Comment (dvargasfuertes) [inline src/features/deepnav.tsx]: we're using `isNavigating` to control whether we do deepnav stuff or not, so seems like we should be able to add/remove this event listener during just the `toggleFeature` method

File: src/features/deepnav.tsx
Comment (dvargasfuertes) [inline src/features/deepnav.tsx]: why did we switch from `document` to `window`?

File: src/features/workBench.ts
Comment (dvargasfuertes) [inline src/features/workBench.ts]: With big files like this, one tactic I like to do is leave a comment like this:

> This is the only substantive change to the file, the rest just add `extensionAPI` as an argument to `addCommand`. Any other differences I flag below

File: src/features/jumpNav.tsx
Comment (dvargasfuertes) [inline src/features/jumpNav.tsx]: Another way to make this readable is to make this more DRY (don't repeat yourself):

```typescript
const commands = [
  {label: "Justify", callback: justify},
  {label: "Toggle Queries", callback: toggleQueries},
  // ... etc.
];

commands.forEach((cmd) => 
  unloads.add(addCommand(cmd, extensionAPI))
);
```

**Review by dvargasfuertes — COMMENTED** [review]
My biggest question is why did we remove all of the hotkeys to the commands?

The goal of the task was to make all of the commands have configurable hotkeys, while still retaining the default value we had so that there's no interruption in functionality

File: src/index.ts
Comment (mdroidian) [reply src/index.ts:179]: Ah, smaller PR's makes sense.

But at the same time, I was discovering it as I went.  So I wouldn't have know that the first PR wouldn't have had to be overwritten / changed.

And when I think:
> What is the smallest PR that we could submit that accomplishes our goal?

If the goal is to migrate all workbench commands to command palette, that first PR wouldn't have achieved that.

I guess I should have spent more time scoping the problem.

File: src/features/dailyNotesPopup.tsx
Comment (mdroidian) [reply src/features/dailyNotesPopup.tsx:638]: RE: 
> My biggest question is why did we remove all of the hotkeys to the commands?
> 
> The goal of the task was to make all of the commands have configurable hotkeys, while still retaining the default value we had so that there's no interruption in functionality

The first reason was that the docs recommend against it, so I was waiting for your thoughts on that.

> Also, the [docs](https://roamresearch.com/#/app/developer-documentation/page/HlEqgAw3H) seem to strongly recommend against setting a default shortcut. Do we still want to go ahead with that?

The second was that I was considering the UX from a new user.  As a new user, lets say I see some features in Workbench that I want to implement in my graph.  I install Workbench that comes preloaded with multiple modules and hotkey shortcuts.  I have to either knowingly turn off the modules I don't want and unset the preconfigured hotkeys or, what seems more likely, unknowingly have some undesired effects down the line.

That being said, I understand causing causing a current user interruptions is not good either.

Neither situation (removing defaults vs setting them), in my opinion, is ideal.  But with the the latter it is a singular occurrence that we can preemptively mitigate by messaging on Twitter / Slack / Email list.  The former is ongoing.

File: src/features/article.tsx
Comment (mdroidian) [reply src/features/article.tsx:329]: Answered in dailyNotesPopup comment

File: src/features/deepnav.tsx
Comment (mdroidian) [reply src/features/deepnav.tsx]: This is additionally poor on my part because the listener is added on `activeDeepNav()` and `toggleFeature()`

So are you saying it is better to have an event listener always on vs only turning it on when needed?

File: src/features/deepnav.tsx
Comment (mdroidian) [reply src/features/deepnav.tsx]: Unintentional mistake.

File: src/features/workBench.ts
Comment (mdroidian) [reply src/features/workBench.ts]: Sorry, I don't follow.
That would be a comment in the code or on the commit?

**Comment by mdroidian** [comment]
> > I went with that because if felt more readable to me than the 200+ lines from the Prettier extension.
> 
> The useful part of the prettier extension is the standardization, not necessarily the decisions it comes with by default. The problem with making exceptions is that someone else working on the team will eventually edit the file, hit save, then prettier will auto format.
> 
> If you disagree with prettier defaults, the scalable solution would be to change the prettier rules, so that it gets formatted automatically. If it's unable to be encoded, then the opinion is subjective, which means it will inevitably be overwritten by somebody else's subjective experience, and now time on the PR is being spent on code formatting instead of code content.

This makes sense, but if someone else were to contribute to this code, how would they be made aware of the prettier / format on save convention?

**Comment by mdroidian** [comment]
Speaking of which, I'm not using the "on save feature".  Should I be?
I just an uneasy feeling when I go to edit a few lines by the commit shows quite a few more changes than I made.

File: src/index.ts
Comment (dvargasfuertes) [reply src/index.ts:179]: Making a note to clarify more on Monday :+1:

File: src/features/dailyNotesPopup.tsx
Comment (dvargasfuertes) [reply src/features/dailyNotesPopup.tsx:638]: So I think it's valid to disregard going forward. But one lesson I've learned is that no amount of messaging will resolve an interruption of behavior.

Particularly ones like `hot_keys`/`jumpnav` which all it does is add more hot keys for the user.

 We are also not precluded from removing it later on if it turns out that too many defaults is not desirable

File: src/features/deepnav.tsx
Comment (dvargasfuertes) [reply src/features/deepnav.tsx]: It's more a comment that `isNavigating` is already doing this work. If we want to do the smart listener approach you suggest (this is probably better, but harder to get right), then we should prob get rid of `isNavigating` since it will add confusion

File: src/features/workBench.ts
Comment (dvargasfuertes) [reply src/features/workBench.ts]: On the commit to make it easier for reviewers looking over big files how to read it going in

**Comment by dvargasfuertes** [comment]
> how would they be made aware of the prettier / format on save convention?

This is somewhat implicit by seeing prettier in a repo in the industry at this point. Sortof like if you see a testing library you can assume that the project is running tests in Github in some capacity and they should all be passing before merging

> Speaking of which, I'm not using the "on save feature". Should I be?

I highly recommend it, it's life changing :+1:

> I just an uneasy feeling when I go to edit a few lines by the commit shows quite a few more changes than I made.

That's okay bc of two reasons:
1. Prettier only does _whitespace_ changes - it does not affect semantic reasoning
2. Because of 1, you can review PRs by checking "hide whitespace changes" to ignore those changes when reviewing

**Comment by mdroidian** [comment]
A few notes / questions on [default hotkeys:](https://github.com/dvargas92495/roamjs-workbench/pull/391/commits/c1d368cbd668aac1a35cdc330ed84dc718c86ed3)

1. DNP Jump Date Forward / Back conflict with the inbuild Roam `Cycle Version`

3. `alt-shift-p` doesn't work for me (in Privacy Mode)

As a matter of fact, I can't set any hotkey using the UI to `alt-shift-p`.
I did check another program, however, and `alt-shift-p` worked, so it doesn't seem to be my system that is the issue.
Could you check if it works for you?

4. How do you want to handle the default hotkeys for Hot Keys (jumpNav.tsx)?

The docs say

> All the hot keys in this module are initiated with a modifier + j + another character. So Alt+j, Ctrl+j and CMD+j are all equivalent in this context. The final character, pressed in succession after modifier + j, is what determines the action taken.

I don't believe we can set more than one hotkey as a default.

5. I believe the code in weekly-notes.ts actually allowed for both hotkeys because `.isIOS` is "true if client is iphone, ipad or ipod"

I coded it to stay in line with the docs

> To jump to this week's page, hit Alt+w or Ctrl+Shift+W (Mac Only).

**Comment by dvargasfuertes** [comment]
> DNP Jump Date Forward / Back conflict with the inbuild Roam Cycle Version

> How do you want to handle the default hotkeys for Hot Keys (jumpNav.tsx)?

Ok so I'm going to double back and what I said before and make an exception for this and jumpnav. Let's remove the defaults, and be sure to make announcements in `#roam-js` and `twitter`, emphasizing that the reason why we removed the defaults here are so that users could configure them in Roam's hot key panel.

> alt-shift-p doesn't work for me (in Privacy Mode)

Works for me - Sometimes chrome extensions could conflict with hot keys

**Comment by mdroidian** [comment]
It might be prudent to not merge this until the docs have been updated.
Also, I have another idea to toss your way on Monday that relates to this.

> Works for me - Sometimes chrome extensions could conflict with hot keys

🤦‍♂️ That was it!

**Review by dvargasfuertes — APPROVED** [review]

**Pull request merged**
